### DISCUSSION TITLE
How does referencing properties and enumerated scalars work?

### DISCUSSION THREAD

**Discussion by astako** [opencaesar/platform#18] [Q&A]
I am creating an embedded vocabulary to map portions of vocab A to vocab B. I would like to create a reference for a scalar property in vocabulary A in the embedded vocabulary to inherit from a scalar property from vocabulary B. The domains of the scalar properties are already mapped.
The range of scalar property A is an enumerated scalar. The range of scalar property B is also an enumerated scalar, but with a different list of enumerations. 

Is there a way to somehow combine the enumerations so that scalarPropertyA can correspond to scalarPropertyB?

I tried the following statements in the embedded vocab: 
ref scalar property A:scalarPropertyA :> B:scalarPropertyB
ref enumerated scalar A:enumsA :> B:enumsB 

The reasoner failed because it found the ontology to be inconsistent. Explanation in reasoning.xml was:  "The literal value does not satisfy the datatype restriction". The enumsB enum scalar was listed in the Datatypes section of reasoning.xml. In the Data properties section the scalarPropertyA was a subclass of scalarPropertyB. It failed on an individual for which scalarPropertyA was asserted with a value from enumsA.

**Comment by melaasar** [comment]
Specifying  `A:scalarPropertyA :> B:scalarPropertyB` means if you asserted `:individual A:scalarPropertyA :value` then the reasoner can infer `:individual B:scalarPropertyB :value`. However, if `:value` is not in the range of property B:scalarPropertyB, then it's an inconsistency.

If what you want to say "for every individual with an assertion on A:scalarPropertyA with literalA (from enumsA), you want the reasoner to infer that the individual has an assertion on B:scalarPropertyB with literalB (from enumsB)", then a better way to do this is by a rule, as follows (no enum or property specialization needed):
```
rule enumMapping [
    A:scalarPropertyA(x, 'literalA') -> B:scalarPropertyB(x, 'literalB')
]
```

**Reply by astako** [reply]
It seems like even though it's technically possible to make references to enumerated scalar properties (according to section 6.11.7. EnumeratedScalarReference of the OML spec) in an embedded vocabulary, for this to work in practice, B:enumB would have to include A:enumA literals. If B:enumB is defined in a completely separate vocabulary, this does not seem practical. 

For the case of mapping/embedding, is the rule method the only way to avoid directly impacting the vocabularies?

**Reply by melaasar** [reply]
There is no requirement that an enum scalar be defined in the same vocabulary as its super enum scalar. For example, the following should be doable::

vocabulary1:
```
   enumerated scalar A [ "One", "Two", "Three" ]
```

vocabulary 2:
```
   enumerated scalar B :> vocabulary1:A
```

The `:>` in the case of enumerated scalars is actually interpreted as equivalency (not specialization). But, I noticed that if I defined literals for `scalar B` that are NOT the same as literals for `scalar A`, the reasoner does not complain! I would have expected it to in the case of enumerated literals, similar to how it does in the case of enumerated individuals.

**Comment by astako** [comment]
We encountered a build issue with referencing scalar properties. Since it is related to the general question of how scalar property referencing is supposed to work, I am including it here.

Using the example of mapping vocabulary A to B via referencing, say we have the following (where ------ denotes separate oml files):

/namespace#vocabA
aspect NamedElement
scalar property Name [
     domain NamedElement
     range string ]
-----------------------------------------------------

/namespace#vocabA
extends /namespace#upperOntology as uo
scalar property hasLocaName :> uo : hasName
-----------------------------------------------------

/namespace#upperOntology
aspect: IdentifiedElement
scalar property hasName [
  domain IdentifyElement
  range string ]

--------------------------------------------------
/namespace#embeddedVocab
extends /namespace#vocabA as va
extends /namespace#vocabB as vb
ref scalar property vb:hasPortLocalName :> vb:Name
-------------------------------------------------------

The above fails to build. The ./gradlew build --stacktrace result is in the attached txt file[
[build-fail.txt](https://github.com/opencaesar/community/files/11142963/build-fail.txt)
](url)

**Reply by melaasar** [reply]
@astako , can you please clarify the example?
- The second vocabulary above is named `vocabA` like the first one.
- The scalar property definition (in 2nd vocab) `hasLocalName` is missing the domain/range
- The `hasPortLocalName` is refed but not defined any where.

**Reply by astako** [reply]
@melaasar , clarification:
Using the example of mapping vocabulary B to A via referencing, say we have the following (where ------ denotes separate oml files):

/namespace#upperOntology
aspect: TitledElement
scalar property hasTitle [
	domain TitledElement
	range string ]

-----------------------------------------------------
/namespace#vocabA
aspect NamedElement
scalar property Name [
	domain NamedElement
	range string ]

-----------------------------------------------------
/namespace#vocabB
extends /namespace#upperOntology as uo
aspect LocalNamed :> uo:TitledElement
scalar property hasLocalName :> uo : hasTitle [
	domain LocalNamed
	range string ]

------------------------------------------------------
/namespace#embeddedVocab
extends /namespace#vocabA as va
extends /namespace#vocabB as vb
ref scalar property vb:hasLocalName :> va:Name
------------------------------------------------------

Above fails on omlToOwl task with message: multiply-rooted property tree

**Reply by melaasar** [reply]
I confirm I can reproduce this. Opening a new bug (https://github.com/opencaesar/owl-adapter/issues/61) on oml2owl adapter

**Reply by melaasar** [reply]
Fixed in 1.4.4 of owl-adapter.
